crates/ruff_linter/src/rules/flake8_type_checking/rules/runtime_import_in_type_checking_block.rs

```diff
@@ -106,74 +106,91 @@ pub(crate) fn runtime_import_in_type_checking_block(
     let mut actions: FxHashMap<(NodeId, Action), Vec<ImportBinding>> = FxHashMap::default();
 
     for binding_id in scope.binding_ids() {
-        let binding = checker.semantic().binding(binding_id);
+        let top_binding = checker.semantic().binding(binding_id);
 
-        let Some(import) = binding.as_any_import() else {
+        let Some(_) = top_binding.as_any_import() else {
             continue;
         };
 
-        let Some(reference_id) = binding.references.first().copied() else {
+        let Some(reference_id) = top_binding.references.first().copied() else {
             continue;
         };
 
-        if binding.context.is_typing()
-            && binding.references().any(|reference_id| {
+        let binding_name = top_binding
+            .name(checker.source())
+            .split('.')
+            .next()
+            .unwrap_or("");
+
+        // NOTE: It’s necessary to go through all bindings, including shadowed ones,
+        //       using the module name.
+        for binding in scope
+            .get_all(&binding_name.nfkc().collect::<String>())
+            .map(|binding_id| checker.semantic().binding(binding_id))
```

Comment:
It might be more efficient to use the existing method on the semantic model to iterate through the shadowed bindings (unless your change makes those bindings not always shadow each other):

https://github.com/astral-sh/ruff/blob/91b7a570c2bd1c9e1cab894ded866e885f28946a/crates/ruff_python_semantic/src/model.rs#L2038-L2075

Reply:
Or rather as pointed out above, you might be able to revert everything and only have to replace `scope.binding_ids()` with `scope.all_bindings()`.